### PR TITLE
Make Proof Rule enum a part of the API

### PR DESCRIPTION
**Pull request by hansjoergschurr**
This pull requests makes the enum that lists all proof rules a part of the API.

It also renames the enum from `PfRule` to `ProofRule`.  It also renames some unrelated types and function names that share the `PfRule` name (such as `DslPfRule`).
This rename unfortunately touches many files since `PfRule` is not an uncommon type.  (second to last commit)

I also added the enum to the Java and Python API, but I am not sure I got this right.  It is in the last commit.

### CODE REVIEW

File: docs/api/java/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,36 @@ if(BUILD_BINDINGS_JAVA)
   set(JAVADOC_OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR}/build/api/java)
   # used to trigger the rebuild
   set(JAVADOC_INDEX_FILE ${JAVADOC_OUTPUT_DIR}/index.html)
+  # HTML added to javadoc to include Mathjax
+  # This is fragile with respect to escapes:
+  #   - The LaTeX sequence '$#1$' must be written as '\$$#1$$'.
+  #   - Whitespaces in some places must be omitted, otherwise
+  #     they become ';'.
+  #   - To make LaTeX commands work they need '\\\\'.
+  set(JAVADOC_MATHJAX [=["
```
Comment (aniemetz) [inline docs/api/java/CMakeLists.txt:31]: How robust is this wrt to future Javadoc and Mathjax versions?

File: include/cvc5/cvc5_proof_rule.h [hunk 2]
```diff
@@ -42,31 +61,34 @@ namespace cvc5::internal {
...
- * (1) :cpp:enumerator:`ASSUME <cvc5::internal::PfRule::ASSUME>`, which represents an open
- * leaf in a proof; and (2) :cpp:enumerator:`SCOPE <cvc5::internal::PfRule::SCOPE>`, which
- * encloses a scope (a subproof) with a set of scoped assumptions. The core rules additionally correspond to
- * generic operations that are done internally on nodes, e.g. calling
- * Rewriter::rewrite.
+ * (1) :cpp:enumerator:`ASSUME <cvc5::ProofRule::ASSUME>`, which represents an open
+ * leaf in a proof; and (2) :cpp:enumerator:`SCOPE <cvc5::ProofRule::SCOPE>`, which
+ * encloses a scope (a subproof) with a set of scoped assumptions. The core rules
+ * additionally correspond to generic operations that are done internally on nodes,
+ * e.g. calling Rewriter::rewrite.
```
Comment (aniemetz) [inline include/cvc5/cvc5_proof_rule.h:85]:
```suggestion
 * e.g., calling Rewriter::rewrite.
```

**Review by aniemetz — APPROVED** [review]

File: include/cvc5/cvc5_proof_rule.h [hunk 3]
```diff
@@ -2306,12 +2329,36 @@ enum class PfRule : uint32_t
...
 
   //================================================= Unknown rule
-  UNKNOWN,
+  EVALUE(UNKNOWN),
 };
+// clang-format on
+
+#ifdef CVC5_API_USE_C_ENUMS
+#ifndef DOXYGEN_SKIP
+typedef enum ENUM(Kind) ENUM(Kind);
```
Comment (ajreynol) [inline include/cvc5/cvc5_proof_rule.h:2341]: Should say ProofRule not Kind?

File: include/cvc5/cvc5_proof_rule.h [hunk 4]
```diff
@@ -2330,14 +2377,25 @@ const char* toString(PfRule id);
  * @param id The proof rule to write to the stream
  * @return The stream
  */
-std::ostream& operator<<(std::ostream& out, PfRule id);
+std::ostream& operator<<(std::ostream& out, ProofRule id);
 
 /** Hash function for proof rules */
-struct PfRuleHashFunction
+struct ProofRuleHashFunction
```
Comment (ajreynol) [inline include/cvc5/cvc5_proof_rule.h:2383]: For consistency, perhaps this should change to the style for the hash function used by Kinds here: https://github.com/cvc5/cvc5/blob/main/include/cvc5/cvc5_kind.h#L5676.

(This code is using the standard practice for internal enums, but now that it is external maybe it should follow that one).

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few requests.  Also, I believe you will need to add the print function for ProofRule to the unit tests here:
https://github.com/cvc5/cvc5/blob/main/test/unit/api/python/test_uncovered.cpp
https://github.com/cvc5/cvc5/blob/main/test/unit/api/java/UncoveredTest.cpp


Also a cpp unit test for printing proof rules needs to be added I think, similar to: 
https://github.com/cvc5/cvc5/blob/main/test/unit/api/cpp/api_kind_black.cpp

**Review by ajreynol — APPROVED** [review]

**Pull request merged**